### ISSUE TITLE
[FEATURE] Is there a way to change the authentication information in the headers when calling the predict API?

### ISSUE THREAD

**Issue by 10000-ki** [opensearch-project/ml-commons#3208]
**Is your feature request related to a problem?**

Currently, we are configuring it to include authentication information when creating a custom ML connector.

```json
POST /_plugins/_ml/connectors/_create
{
  "name": "Custom Connector: embeddings",
  "description": "The connector to Custom X",
  "version": 1,
  "protocol": "http",
  "credential": {
      "auth": "${token}"
  },
  "parameters": {
      "endpoint": "${endpoint}"
  },
  "actions": [
        {
          "action_type": "predict",
          "method": "POST",
          "url": "https://${parameters.endpoint}",
          "headers": {
            "Authorization": "Bearer ${credential.auth}",
            "Content-Type": "application/json"
          },
          "request_body": "{ \"text\": \"${parameters.inputText}\" }",
        }
    ]
}
```

`The issue is that this authentication token has an expiration period, so it needs to be renewed periodically.`

**What solution would you like?**

Is there a way to pass the auth token value again at the point of calling the predict API or the custom connector API?

```json
POST /_plugins/_ml/models/${model_id}/_predict
{
    "parameters": {
        "inputText": "text embedding..."
    }
}
```

**What alternatives have you considered?**

**Do you have any additional context?**

**Comment by brianf-aws** [comment]
Actually this is a better alternative instead of un-deploying you can do it like how it states here https://opensearch.org/docs/latest/ml-commons-plugin/remote-models/connectors/#updating-connector-credentials

**Comment by ylwu-amzn** [comment]
@b4sjoo , I think we have some limitation for updating connector credential. Can you clarify ?

**Comment by 10000-ki** [comment]
@brianf-aws 

> Actually this is a better alternative instead of un-deploying you can do it like how it states here https://opensearch.org/docs/latest/ml-commons-plugin/remote-models/connectors/#updating-connector-credentials

oh thx!!! i will try

**Comment by b4sjoo** [comment]
@10000-ki Yeah so currently our stand-alone connector does not support update on-the-fly, right now the work around is to use the inner connector in that link. We have the plan to support the stand-alone connector in the future, but we haven't come into a detailed design yet.

**Comment by dhrubo-os** [comment]
@10000-ki Does that unblock your use-case?

**Comment by 10000-ki** [comment]
@dhrubo-os 

> Does that unblock your use-case?

No, [the API](https://opensearch.org/docs/latest/ml-commons-plugin/remote-models/connectors/#updating-connector-credentials) is not functioning properly. The credentials are not updated correctly. It seems there might be an issue with the API, i think we need to review the code.

**Comment by dhrubo-os** [comment]
@b4sjoo could you please check?

**Comment by 10000-ki** [comment]
@brianf-aws 

When updating the connector, the update fails with the following error.

Is there something I might have missed?


```
PUT /_plugins/_ml/models/Fsp2BZMBO39bHfNca1lv
{
  "connector": {
    "credential": {
      "auth": "..."
    }
  }
}
```

```
{
  "error": {
    "root_cause": [
      {
        "type": "null_pointer_exception",
        "reason": "Cannot invoke \"org.opensearch.ml.common.connector.Connector.update(org.opensearch.ml.common.transport.connector.MLCreateConnectorInput, java.util.function.Function)\" because \"connector\" is null"
      }
    ],
    "type": "null_pointer_exception",
    "reason": "Cannot invoke \"org.opensearch.ml.common.connector.Connector.update(org.opensearch.ml.common.transport.connector.MLCreateConnectorInput, java.util.function.Function)\" because \"connector\" is null"
  },
  "status": 500
}
```

**Comment by 10000-ki** [comment]
![Image](https://github.com/user-attachments/assets/8afd1e3c-14d7-4082-a981-72d84984dc86)

**Comment by brianf-aws** [comment]
Hey @10000-ki, apologies for the late response. I just tried updating a connector myself and it seems to work. I am rechecking your post and see there is a minor typo

```
PUT /_plugins/_ml/models/Fsp2BZMBO39bHfNca1lv
{
  "connector": { <--- lets change this to connectors
    "credential": {
      "auth": "..."
    }
  }
}
```

please see below for proof (my local)

![Image](https://github.com/user-attachments/assets/c3d3858c-293d-4985-b72c-a8d9204c76e1)

Please reach out if you are facing any other blockers!

**Comment by b4sjoo** [comment]
@brianf-aws This does not seem quite right to me... I just checked the code and it's in single form

**Comment by b4sjoo** [comment]
@10000-ki it seems you have used the connector creation API to create this connector, so I supposed you are using the stand-alone connector?

**Comment by brianf-aws** [comment]
Hey @10000-ki, thanks to @b4sjoo we reviewed internally and noticed that there is a bug when dealing with standalone connectors.

For now you might have to undeploy and make the model use a inline connector please follow an example [here](https://opensearch.org/docs/latest/ml-commons-plugin/api/model-apis/register-model/#example-request-externally-hosted-with-a-connector-specified-as-part-of-the-model)

Now subsequent updates like this
```
PUT {{ _.domain }}/_plugins/_ml/models/{{ _.model_id }}
{
	"connector" : {
		"parameters": {
        "endpoint": "api.openai.com",
        "model": "gpt-3.5-turb110"
    }
	}
}
# I tested and was able to update the connector correctly
```

I will create a issue so we can track this bug.

**Comment by b4sjoo** [comment]
Besides if you want to update the standalone connector itself:
```
PUT {{ _.domain }}/_plugins/_ml/connectors/{{ _.connector_id }}
{
    "credential": {
      "auth": "..."
    }
}
```

**Comment by 10000-ki** [comment]
> Besides if you want to update the standalone connector itself:

yes 

I am using a custom model connector, and currently, I stop, modify, and restart the model to update the connection information.
It would be great if the bug were fixed so that changes could be made without downtime.

**Comment by brianf-aws** [comment]
hi @10000-ki I can understand why this is a bad experience for you.

Since the standalone connector makes you have to undeploy, update connector and deploy. I think for a solution you can create a connector within a model ([link](https://opensearch.org/docs/latest/ml-commons-plugin/api/model-apis/register-model/#example-request-externally-hosted-with-a-connector-specified-as-part-of-the-model)). And then perform updates on it ([doc](https://opensearch.org/docs/latest/ml-commons-plugin/api/model-apis/update-model/#:~:text=Contains%20specifications%20for%20a%20connector%20for%20a%20model%20hosted%20on%20a%20third%2Dparty%20platform.%20For%20more%20information%2C%20see%20Creating%20a%20connector%20for%20a%20specific%20model.%20For%20information%20about%20the%20updatable%20fields%20within%20a%20connector%2C%20see%20Update%20Connector%20API%20request%20fields.))  This will be like how you have it [here](https://github.com/opensearch-project/ml-commons/issues/3208#issuecomment-2599599956).

**Comment by ylwu-amzn** [comment]
> hi [@10000-ki](https://github.com/10000-ki) I can understand why this is a bad experience for you.
> 
> Since the standalone connector makes you have to undeploy, update connector and deploy. I think for a solution you can create a connector within a model ([link](https://opensearch.org/docs/latest/ml-commons-plugin/api/model-apis/register-model/#example-request-externally-hosted-with-a-connector-specified-as-part-of-the-model)). And then perform updates on it ([doc](https://opensearch.org/docs/latest/ml-commons-plugin/api/model-apis/update-model/#:~:text=Contains%20specifications%20for%20a%20connector%20for%20a%20model%20hosted%20on%20a%20third%2Dparty%20platform.%20For%20more%20information%2C%20see%20Creating%20a%20connector%20for%20a%20specific%20model.%20For%20information%20about%20the%20updatable%20fields%20within%20a%20connector%2C%20see%20Update%20Connector%20API%20request%20fields.)) This will be like how you have it [here](https://github.com/opensearch-project/ml-commons/issues/3208#issuecomment-2599599956).

Thanks @brianf-aws. 
@10000-ki, It will take some time to support updating standalone connector on the fly.  Can you try this solution to unblock your use case first?

**Comment by 10000-ki** [comment]
> I think for a solution you can create a connector within a model ([link](https://opensearch.org/docs/latest/ml-commons-plugin/api/model-apis/register-model/#example-request-externally-hosted-with-a-connector-specified-as-part-of-the-model))

@brianf-aws thank u!! i will try

**Comment by 10000-ki** [comment]
> It will take some time to support updating standalone connector on the fly. Can you try this solution to unblock your use case first?

@ylwu-amzn 

No, it’s not urgent at the moment. I’ll try the suggested method first.
Thank you!
